### PR TITLE
Allow to pass chargeback relations by id or href in rates collection

### PR DESCRIPTION
**Pull request by lpichler**
This PR extends `api/rates` endpoint.

- currency, chargeable field can be specified by id or href in request
- chargeback tiers can specified by hash with values:
-  chargeable_fields collections was added just GET  `app/chargeable_fields` and `app/chargeable_fields/1`

**Create**

```


POST api/rates
{
  "per_time" : "daily",
  "chargeback_rate_id" : "1",
  "chargeable_field_id" : "1",
  "description": "My CPU allocation rate",
  "group" : "cpu",
  "per_unit" : "megahertz",	
  "source" : "allocated",
  "chargeback_rate_detail_measure_id": 1,
  "detail_currency" : {"id": 5},
   "chargeable_field" : {"href": "http://localhost:3090/api/chargeable_fields/3"},
  "chargeback_tiers" : [
  	{
	  "start": 0.0,
	  "finish":1,
	  "fixed_rate": 2.0,
	  "variable_rate": 0.0
  	},
  	{
	  "start": 1,
	  "finish": "infinity",
	  "fixed_rate": 2.0,
	  "variable_rate": 0.0
  	}
  ]
}


Response:

{
    "results": [
        {
            "href": "http://localhost:3090/api/rates/1859",
            "id": "1859",
            "enabled": true,
            "description": "My CPU allocation rate",
            "group": "cpu",
            "source": "allocated",
            "metric": null,
            "per_time": "daily",
            "per_unit": "megahertz",
            "friendly_rate": "Daily @ 2.0 + 0.0 per Megahertz from 0.0 to 1.0\nDaily @ 2.0 + 0.0 per Megahertz from 1.0 to Infinity",
            "chargeback_rate_id": "1",
            "created_on": "2020-05-26T10:34:39Z",
            "updated_on": "2020-05-26T10:34:39Z",
            "chargeback_rate_detail_measure_id": "1",
            "chargeback_rate_detail_currency_id": "5",
            "chargeable_field_id": "3",
            "sub_metric": null
        }
    ]
}


```

**Edit**

```


POST /api/rates/1800
{
  "action" : "edit",
  "resource" : { 
  	"detail_currency" : {"id": 5}, 
  	"detail_measure" : {"id": 1},
  	"chargeback_tiers" : [
  		{
		  "start": 0.0,
		  "finish": "infinity",
		  "fixed_rate": 2.0,
		  "variable_rate": 0.0
  		}
  	]
  }
}

Response :
{
    "href": "http://localhost:3090/api/rates/1800",
    "id": "1800",
    "chargeback_rate_detail_currency_id": "5",
    "chargeback_rate_detail_measure_id": "1",
    "chargeback_rate_id": "1",
    "chargeable_field_id": "3",
    "enabled": true,
    "description": "Allocated CPU Count",
    "group": "net_io",
    "source": null,
    "metric": null,
    "per_time": "hourly",
    "per_unit": "cpu",
    "friendly_rate": "Hourly @ 2.0 + 0.0 per Cpu from 0.0 to Infinity",
    "created_on": "2020-04-23T15:00:29Z",
    "updated_on": "2020-05-26T11:44:39Z",
    "sub_metric": null
}



```

@miq-bot enhancement 

@miq-bot assign @abellotti 

cc @gtanzillo

### CODE REVIEW

**Comment by lpichler** [comment]
@miq-bot add_label jansa/yes

File: app/controllers/api/rates_controller.rb [hunk 1]
```diff
@@ -1,9 +1,59 @@
...
+
+    def parse_tiers(data)
+      raise BadRequestError, "chargeback_tiers needs to be specified" if data['chargeback_tiers'].nil?
+
+      tiers = data['chargeback_tiers'].map do |tier_parameters|
+        if (tier_parameters['start'] && tier_parameters['finish']).nil?
+          raise BadRequestError, "Attributes start and finish have to be specified for chargeback tier."
+        end
+
+        tier_parameters['finish'] = Float::INFINITY if tier_parameters['finish'] == "infinity"
```
Comment (abellotti) [inline app/controllers/api/rates_controller.rb:42]: maybe `if tier_parameters["finish"].downcase == "infinity"`

Comment (abellotti) [inline app/controllers/api/rates_controller.rb:42]: Maybe `if tier_parameters["finish"].downcase == "infinity"`

File: app/controllers/api/rates_controller.rb [hunk 2]
```diff
@@ -1,9 +1,59 @@
...
+
+      super
+    end
+
+    def parse_relation_from(data, parameter_key, type, collection)
+      resource_relation = data[parameter_key]
+
+      if resource_relation.present?
+        resource_id = parse_id(resource_relation, collection)
+        raise BadRequestError, "Missing #{type} identifier href or id" unless resource_id
```
Comment (abellotti) [inline app/controllers/api/rates_controller.rb:27]: maybe add test cases for these errors

File: spec/requests/rates_spec.rb [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+          :source           => "used",
+          :enabled          => true
+        }
+        post(api_rates_url, :params => chargeback_params)
+      end.to change(ChargebackRateDetail, :count).by(1)
+
+      expect_result_to_match_hash(response.parsed_body["results"].first, "description" => "rate_#{index}", "enabled" => true)
+      expect(response.parsed_body["results"].first['chargeback_rate_detail_currency_id'].to_i).to eq(currency.id)
+      expect(response.parsed_body["results"].first['chargeable_field_id'].to_i).to eq(chargeable_field.id)
+      expect(response.parsed_body["results"].first['chargeback_rate_id'].to_i).to eq(chargeback_rate.id)
```
Comment (abellotti) [inline spec/requests/rates_spec.rb:46]: maybe replace this with a single expect

File: spec/requests/rates_spec.rb [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+
+    parameters_relations.each do |relations|
+      api_basic_authorize action_identifier(:rates, :edit)
+      relations[:description] = "rate_1"
+      post api_rate_url(nil, chargeback_rate_detail), :params => gen_request(:edit, relations)
+
+      expect(response.parsed_body["description"]).to eq("rate_1")
+      expect(response.parsed_body["chargeback_rate_detail_currency_id"].to_i).to eq(currency_2.id)
+      expect(response.parsed_body['chargeable_field_id'].to_i).to eq(chargeable_field.id)
+      expect(response.parsed_body['chargeback_rate_id'].to_i).to eq(chargeback_rate.id)
```
Comment (abellotti) [inline spec/requests/rates_spec.rb:93]: same as earlier, replace 90-93 with single expect

File: spec/requests/rates_spec.rb [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+      expect(response).to have_http_status(:ok)
+      expect(chargeback_rate_detail.reload.description).to eq("rate_1")
+
+      tiers = chargeback_rate_detail.reload.chargeback_tiers.sort_by(&:start)
+      tiers.zip(parameter_tiers).each do |tier, tier_parameters_original|
+        tier_parameters = tier_parameters_original.dup
+
+        tier_attributes = tier.attributes.except('id')
+        tier_parameters['chargeback_rate_detail_id'] = response.parsed_body['id'].to_i
+        tier_parameters['finish'] = Float::INFINITY if tier_parameters['finish'] == 'infinity'
```
Comment (abellotti) [inline spec/requests/rates_spec.rb:104]: just curious, I see this pattern from before, can this logic be put in a method ?

**Comment by lpichler** [comment]
@abellotti I updated PR according to your suggestions. Thanks for the review 👍

**Comment by abellotti** [comment]
LGTM!!   Thanks @lpichler for the API Enhancement and updates. 👍

**Comment by simaishi** [comment]
Jansa backport details:

```
$ git log -1
commit 12d2fcb9ea506e865b23a6fbc098fa8d6ddf0904
Author: Alberto Bellotti <abellotti@users.noreply.github.com>
Date:   Tue Jun 2 09:40:04 2020 -0400

    Merge pull request #847 from lpichler/pass_relations_to_rates_collection

    Allow to pass chargeback relations by id or href in rates collection

    (cherry picked from commit 9702e4056d7e3469a3305abcd4598ca6451105c7)
```

**Pull request merged**